### PR TITLE
New Mid Layer for serialization

### PR DESCRIPTION
**Pull request by houseroad**
To deentangle proto from serialization and deserialization, we will introduce the a new mid layer (light weight) to store the model in memory.

### CODE REVIEW

File: caffe2/serialize/intermediate_model.h [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+
+#include <ATen/core/Allocator.h>
+
+#include "caffe2/proto/caffe2_pb.h"
+#include "caffe2/proto/torch_pb.h"
+#include "caffe2/serialize/inline_container.h"
+
+namespace at {
+
+struct IntermediateDeviceOption {
```
Comment (dzhulgakov) [inline caffe2/serialize/intermediate_model.h]: first.. you use it only in Tensor. If you actually check which fields would be later used it's actually only deviceType + deviceId - the rest are the artifact of reusing DeviceOption (what does random seed for tensor mean?). So we can probably nuke them from this struct.

second.. it's less clear whether persisting the device info in the model is preferred or not. C2 right now has option keep_device (https://github.com/pytorch/pytorch/blob/master/caffe2/operators/load_save_op.h#L84) which is default false but sometime is true. PyTorch has the default of preserving the device and allowing to turn it off (https://discuss.pytorch.org/t/on-a-cpu-device-how-to-load-checkpoint-saved-on-gpu-device/349). But I guess that storing this info is useful

File: caffe2/serialize/intermediate_model.h [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+  // update functions
+  void update(caffe2::TensorProto* tensor_proto,
+      std::unordered_map<int64_t, at::DataPtr>* id_tensor) {
+    CAFFE_ENFORCE(tensor_proto->has_data_type(), "no data_type in TensorProto!");
+    dataType_ = tensor_proto->data_type();
+    for (int i = 0; i < tensor_proto->dims_size()) {
+      dims_.push_back(tensor_proto->dims(i));
+    }
+    CAFFE_ENFORCE(tensor_proto->has_name(), "no name in TensorProto!");
+    name_ = tensor_proto->name();
```
Comment (dzhulgakov) [inline caffe2/serialize/intermediate_model.h]: do we actually record anything meaningful in `name` in this format?

File: caffe2/serialize/intermediate_model.h [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+        CAFFE_ENFORCE(external_data->has_source_type(), "no source_type in TensorProto!");
+        int64_t source_type = external_data->source_type();
+        if (source_type == ExternalDataProto_SourceType::ExternalDataProto_SourceType_INLINE_CONTAINER) {
+          CAFFE_ENFORCE(external_data->has_record_id(), "no record_id in ExternalDataProto!");
+          int64_t record_id = external_data->record_id();
+          auto it = id_tensor.find(record_id);
+          if (it == id_tensor.end()) {
+            CAFFE_THROW("Tensor's data is missing, tensor name is ", name_);
+          }
+          dataPtr_ = std::move(it->second);
```
Comment (dzhulgakov) [inline caffe2/serialize/intermediate_model.h]: how does it work with shared storages? you'd move out the storage first time (not sure whether it was tested or not in the prototype, but we should be able to serialize aliasing of storages when several tensors point to the same storage and we serialize storage once)

File: caffe2/serialize/intermediate_model.h [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+  bool noContent_ = false;
+};
+
+class IntermediateParameter final {
+ public:
+  // constructor
+  explicit IntermediateParameter(torch::ParameterDef* param_def,
+      std::unordered_map<std::string, IntermediateTensor*>* id_tensor) {
+    if (param_def->has_is_buffer()) {
+      isBuffer_ = param_def->is_buffer();
```
Comment (dzhulgakov) [inline caffe2/serialize/intermediate_model.h]: nit: just reading the field is fine

**Review by dzhulgakov — COMMENTED** [review]
this code is not finished / not tested, thus leaving just a few comments

File: caffe2/serialize/intermediate_model.h [hunk 5]
```diff
@@ -0,0 +1,600 @@
...
+void serializeIntermediateModel(IntermediateModel* imodel, const std::string& filename) {
+  torch::jit::PyTorchFileWriter writer(filename);
+  serializeIntermediateModel(imodel, &writer);
+  writer.writeEndOfFile();
+}
+
+// deserialize an IntermediateModel through a reader,
+// serialize tensors' data first, and maintain the mappint from
+// record id to tensor data
+void deserializeIntermediateModel(IntermediateModel* imodel,
```
Comment (zdevito) [inline caffe2/serialize/intermediate_model.h]: We took great care writing the original deserializer so that it would not require you to read through all the tensor entries and their memory before reading the model so that loading the model can be fast, and that tensor data could be loaded later by mmapping the data sections of the file. It seems like this approach doesn't support that because it reads through the sections eagerly. Can we just load a tensor when we find it in the model? If the file  was mmap it would be possible to initialize the Tensor without ever reading the Tensor records.

File: caffe2/serialize/intermediate_model.h [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+
+// multiple tensor may share the same content
+// SharedData contains:
+//    1) record id (i.e., the offset in the inline container)
+//    2) size, the size of the content
+//    3) data, in serialize, IntermediateModel does NOT own the data,
+//       in deserialize, the data pointer is returned by PyTorchFileReader,
+//       and IntermediateModel owns the data. The ownership later will be
+//       transferred to Tensor
+class SharedData {
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/serialize/intermediate_model.h]: yeah, will do it

File: caffe2/serialize/intermediate_model.h [hunk 7]
```diff
@@ -0,0 +1,600 @@
...
+  uint64_t recordId_;
+  at::DataPtr dataPtr_;
+  uint64_t size_;
+};
+
+// IntermediateDeviceOption stores device related information
+struct IntermediateDeviceOption {
+  int32_t deviceType = 0;
+  int32_t deviceId;
+  bool hasDeviceId = false;
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/serialize/intermediate_model.h]: good idea

File: caffe2/serialize/intermediate_model.h [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+//   2) shape information
+//   3) pointer to the data (including offset and strides)
+class IntermediateTensor final {
+ public:
+  // constructor
+  IntermediateTensor() = default;
+
+  // extract data from TensorProto, called in deserialize
+  // assume record id to data mapping is complete
+  void update(caffe2::TensorProto* tensor_proto,
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/serialize/intermediate_model.h]: yeah, will remove most implementation to cpp, and make sure, we don't include protobuf headers in this header.

File: caffe2/serialize/intermediate_model.h [hunk 9]
```diff
@@ -0,0 +1,600 @@
...
+          }
+          break;
+        }
+      case caffe2::TensorProto_StorageType_NO_CONTENT:
+        {
+          noContent_ = true;
+          break;
+        }
+      default:
+        AT_ERROR("Uknown storage_type %lld", storage_type);
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/serialize/intermediate_model.h]: oh, I remember AT_ASSERTM was formatted string... my memory was bad...

File: caffe2/serialize/intermediate_model.h [hunk 10]
```diff
@@ -0,0 +1,600 @@
...
+    for (auto dim : dims_) {
+      tensor_proto->add_dims(dim);
+    }
+    tensor_proto->set_data_type(static_cast<caffe2::TensorProto_DataType>(dataType_));
+    // NB: maybe later we support RAW
+    tensor_proto->set_storage_type(caffe2::TensorProto_StorageType::TensorProto_StorageType_EXTERNAL);
+    caffe2::ExternalDataProto* data_proto = tensor_proto->mutable_external_data();
+    // NB: maybe later we support SIMPLE_FILE
+    data_proto->set_source_type(caffe2::ExternalDataProto_SourceType_INLINE_CONTAINER);
+    data_proto->set_record_id(std::to_string(data_->recordId()));
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/serialize/intermediate_model.h]: yeah, we should

File: caffe2/serialize/intermediate_model.h [hunk 11]
```diff
@@ -0,0 +1,600 @@
...
+// record id to tensor data
+void deserializeIntermediateModel(IntermediateModel* imodel,
+    torch::jit::PyTorchFileReader* reader) {
+  std::unordered_map<uint64_t, std::shared_ptr<SharedData>> id_data;
+  while (reader->hasNextRecord()) {
+    at::DataPtr data_ptr;
+    size_t data_key;
+    size_t data_size;
+    std::tie(data_ptr, data_key, data_size) = reader->getNextRecord();
+    if (reader->hasNextRecord()) {
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/serialize/intermediate_model.h]: ok, will do

File: caffe2/serialize/intermediate_model.h [hunk 12]
```diff
@@ -0,0 +1,600 @@
...
+
+  explicit IntermediateParameter(torch::ParameterDef* param_def,
+      std::unordered_map<uint64_t, std::shared_ptr<SharedData>>* id_data) {
+    AT_ASSERTM(param_def->has_name(), "ParameterDef has no name! %s",
+        param_def->DebugString());
+    name_ = param_def->name();
+    isBuffer_ = param_def->is_buffer();
+    requireGradient_ = param_def->require_gradient();
+    if (param_def->has_tensor()) {
+      tensor_.update(param_def->mutable_tensor(), id_data);
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/serialize/intermediate_model.h]: This is prepared for the old embeded weights in TensorProto

File: caffe2/serialize/intermediate_model.h (hunk 5 above)
Comment (houseroad) [reply caffe2/serialize/intermediate_model.h]: Yes, I will add a lazy mode in deserialize, which only load the model metadata, and we can feed the data to the tensor later. Right now we keep the model as the last record, and corresponding functionality in the FileReader.

File: caffe2/serialize/intermediate_model.h [hunk 13]
```diff
@@ -0,0 +1,600 @@
...
+  std::unordered_map<uint64_t, std::shared_ptr<SharedData>> id_data;
+  while (reader->hasNextRecord()) {
+    at::DataPtr data_ptr;
+    size_t data_key;
+    size_t data_size;
+    std::tie(data_ptr, data_key, data_size) = reader->getNextRecord();
+    if (reader->hasNextRecord()) {
+      auto it = id_data.find(data_key);
+      if (it != id_data.end()) {
+        it->second->setDataPtr(std::move(data_ptr));
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/serialize/intermediate_model.h]: Good catch, we should throw error here.

File: caffe2/serialize/intermediate_model.cc [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+  if (method_def->has_torch_script()) {
+    torchScript_ = method_def->torch_script();
+  } else if (method_def->has_graph()) {
+    graph_.reset(method_def->release_graph());
+  } else {
+    AT_ERROR("No method body is found!");
+  }
+}
+
+IntermediateMethod::IntermediateMethod(const IntermediateMethod& method) {
```
Comment (dzhulgakov) [inline caffe2/serialize/intermediate_model.cc]: why do you ever need to have copy / assignment constructors for these?

File: caffe2/serialize/intermediate_model.cc [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+        dataptr->recordId = it->second;
+      } else {
+        uint64_t id = writer->writeRecord(data, size);
+        dataptr->recordId = id;
+        data_id[data] = id;
+      }
+    }
+
+    auto* subms = m->mutableSubmodules();
+    for (int i = 0; i < subms->size(); ++i) {
```
Comment (dzhulgakov) [inline caffe2/serialize/intermediate_model.cc]: range for?

File: caffe2/serialize/intermediate_model.cc [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+void serializeIntermediateModel(IntermediateModel* imodel,
+    torch::jit::PyTorchFileWriter* writer) {
+  std::unordered_map<void*, uint64_t> data_id;
+  std::stack<IntermediateModule*> imodules;
+  imodules.push(imodel->mutableMainModule());
+  while (!imodules.empty()) {
+    IntermediateModule* m = imodules.top();
+    imodules.pop();
+    auto* params = m->mutableParameters();
+    for (int i = 0; i < params->size(); ++i) {
```
Comment (dzhulgakov) [inline caffe2/serialize/intermediate_model.cc]: range_for?

File: caffe2/serialize/intermediate_model.cc [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+void serializeIntermediateModel(IntermediateModel* imodel, const std::string& filename) {
+  torch::jit::PyTorchFileWriter writer(filename);
+  serializeIntermediateModel(imodel, &writer);
+  writer.writeEndOfFile();
+}
+
+void deserializeIntermediateModel(IntermediateModel* imodel,
+    torch::jit::PyTorchFileReader* reader, DeserializeMode mode) {
+  std::unordered_map<uint64_t, std::shared_ptr<SharedData>> id_data;
+  if (mode == DeserializeMode::LAZY) {
```
Comment (dzhulgakov) [inline caffe2/serialize/intermediate_model.cc]: LAZY/EAGER are too easy to mix up with run modes. How about LOAD_TENSOR_DATA vs HEADER_ONLY?

File: caffe2/serialize/intermediate_model.cc [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+  AT_ASSERT(mode == DeserializeMode::EAGER);
+  while (reader->hasNextRecord()) {
+    at::DataPtr data_ptr;
+    size_t data_key;
+    size_t data_size;
+    std::tie(data_ptr, data_key, data_size) = reader->getNextRecord();
+    if (!reader->hasNextRecord()) {
+      // the last record is model data (ModelDef)
+      torch::ModelDef model_def = torch::ModelDef();
+      model_def.ParsePartialFromArray(data_ptr.get(), data_size);
```
Comment (dzhulgakov) [inline caffe2/serialize/intermediate_model.cc]: I think you need to check the result value. And why Partial version?

File: caffe2/serialize/intermediate_model.cc [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+      imodel->update(&model_def, &id_data, mode);
+      continue;
```
Comment (dzhulgakov) [inline caffe2/serialize/intermediate_model.cc]: it's more readable with `break` here

File: caffe2/serialize/intermediate_model.h [hunk 20]
```diff
@@ -0,0 +1,385 @@
...
+  explicit SharedData(uint64_t record_id) : recordId(record_id), dataPtr() {}
+  explicit SharedData(uint64_t record_id, at::DataPtr&& data_ptr, uint64_t size)
+    : recordId(record_id), dataPtr(std::move(data_ptr)), size(size) {}
+
+  c10::optional<uint64_t> recordId;
+  at::DataPtr dataPtr;
+  // For deserialize:
+  //   1) in EAGER mode, size information is avaiable immediately
+  //   2) in LAZY mode, until we load the tensor data, size won't be available
+  uint64_t size = 0;
```
Comment (dzhulgakov) [inline caffe2/serialize/intermediate_model.h]: that's actually a minor nuisance for future mmap - we don't have the size of the shared storage without seeking to the record. We could store the storage size in the metadata protobuf as well. But I'm not sure it's worth doing now.

Comment (zdevito) [reply caffe2/serialize/intermediate_model.h]: Size information should be part of the metadata file, always. We don't want to have to jump around the file just to setup tensors. Furthermore, adding a single meta-data field to the records themselves will encourage more people to stash things with the records. Eventually this will devolve into having some fields next to the records and some in the meta-data file, which would be a very bad situation for versioning. My opinion is that all records should only be black boxes and that understanding them will require reading the meta-data file to find out anything else.

Comment (houseroad) [reply caffe2/serialize/intermediate_model.h]: Sure, I will add the size field to the metadata (i.e., protobuf) as well.

File: caffe2/serialize/intermediate_model.cc (hunk 14 above)
Comment (houseroad) [reply caffe2/serialize/intermediate_model.cc]: because windows build ask for `IntermediateMethod& IntermediateMethod::operator =(const IntermediateMethod& method)`, and if I add it, linux started asking for `IntermediateMethod::IntermediateMethod(const IntermediateMethod& method) `

Comment (houseroad) [reply caffe2/serialize/intermediate_model.cc]: I think by default it's deleted on Windows

File: caffe2/serialize/intermediate_model.cc (hunk 15 above)
Comment (houseroad) [reply caffe2/serialize/intermediate_model.cc]: Sure

File: caffe2/serialize/intermediate_model.cc (hunk 16 above)
Comment (houseroad) [reply caffe2/serialize/intermediate_model.cc]: Sure

File: caffe2/serialize/intermediate_model.cc (hunk 17 above)
Comment (houseroad) [reply caffe2/serialize/intermediate_model.cc]: Sounds great!

File: caffe2/serialize/intermediate_model.cc (hunk 18 above)
Comment (houseroad) [reply caffe2/serialize/intermediate_model.cc]: Agree ParseFromArray is more restricted, we should use it and ASSERT on the results.

File: caffe2/serialize/intermediate_model.cc (hunk 19 above)
Comment (houseroad) [reply caffe2/serialize/intermediate_model.cc]: Sure

File: caffe2/serialize/intermediate_model.cc (hunk 14 above)
Comment (dzhulgakov) [reply caffe2/serialize/intermediate_model.cc]: wait, why does windows build require a copy constructor? it's probably because you stash things into a vector. So you could add a move constructor (marked noexcept) and then delete the copy constructor. (or the lazy solution - store a vector of unique pointers)

There shouldn't be a place where you copy the IntermediateMethod, really.

File: caffe2/serialize/intermediate_model.cc [hunk 21]
```diff
@@ -0,0 +1,335 @@
...
+void serializeIntermediateModel(IntermediateModel* imodel,
+    torch::jit::PyTorchFileWriter* writer) {
+  std::unordered_map<void*, uint64_t> data_id;
+  std::stack<IntermediateModule*> imodules;
+  imodules.push(imodel->mutableMainModule());
+  while (!imodules.empty()) {
+    IntermediateModule* m = imodules.top();
+    imodules.pop();
+    for (auto& param : *(m->mutableParameters())) {
+      std::shared_ptr<SharedData> dataptr = param.mutableTensor()->data();
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/serialize/intermediate_model.cc]: Yep, you will see that in my next diff :-) https://github.com/pytorch/pytorch/pull/13587/files#diff-68bb091a9b085d615b756851425fe6b8R132

File: caffe2/proto/caffe2.proto [hunk 22]
```diff
@@ -31,9 +31,11 @@ message ExternalDataProto {
   optional string record_id = 2;
   // the offset of the starting point, the content may be shared between
   // multiple tensors
-  optional int64 offset = 3;
+  optional int64 offset = 3 [default = 0];
   // the strides of the content
   repeated int64 strides = 4;
+  // the size of the record
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/proto/caffe2.proto]: Yeah, I will change it and make its meaning clear.

File: caffe2/serialize/intermediate_model_test.cc [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+  std::shared_ptr<serialize::SharedData> data =
+    std::make_shared<serialize::SharedData>(0, raw_size, std::move(data_ptr));
+  tensor1->setData(data);
+  tensor1->setStrides(strides);
+  // prepare second parameter, share the data with first parameter
+  std::string param2_name = "Test-Parameter-2-Name";
+  params->emplace_back(param2_name, is_buffer, require_gradient);
+  serialize::IntermediateTensor* tensor2 = params->at(1).mutableTensor();
+  *tensor2 = serialize::IntermediateTensor(caffe2::TensorProto_DataType_FLOAT, dims, 0);
+  tensor2->setData(data);
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/serialize/intermediate_model_test.cc]: Will add it in this diff.

**Review by jamesr66a — APPROVED** [review]
Generally looks fine, but do you have a description of what the overall design will look like?

File: caffe2/serialize/intermediate_model.h [hunk 24]
```diff
@@ -0,0 +1,385 @@
...
+  class ParameterDef;
+  class MethodDef;
+  class ModuleDef;
+  class ModelDef;
+}
+
+namespace at {
+namespace serialize {
+
+enum DeserializeMode {
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/serialize/intermediate_model.h]: Good point, changed it.

File: caffe2/serialize/intermediate_model.cc [hunk 25]
```diff
@@ -0,0 +1,335 @@
...
+            // tensor data is only loaded in LOADER_TENSOR_DATA mode
+            if (it == id_data->end()) {
+              AT_ERROR("Tensor's data is missing in id_data, tensor name is ",
+                  name_, ", and record_id is ", caffe2::to_string(record_id));
+            }
+            data_ = it->second;
+            AT_ASSERT(data_->recordId.value() == record_id);
+            AT_ASSERT(data_->size == record_size);
+          } else {
+            AT_ASSERTM(mode == DeserializeMode::HEADER_ONLY, "unkonw deserialize mode.");
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/serialize/intermediate_model.cc]: good catch~

File: caffe2/serialize/intermediate_model.cc [hunk 26]
```diff
@@ -0,0 +1,335 @@
...
+void IntermediateParameter::dump(torch::ParameterDef* param_def) {
+  param_def->set_name(name_);
+  param_def->set_is_buffer(isBuffer_);
+  param_def->set_require_gradient(requireGradient_);
+  caffe2::TensorProto* tensor_def = param_def->mutable_tensor();
+  tensor_.dump(tensor_def);
+}
+
+IntermediateMethod::IntermediateMethod(torch::MethodDef* method_def) {
+  //AT_ASSERTM(method_def->has_name(), "name is required for MethodDef!");
```
Comment (dzhulgakov) [inline caffe2/serialize/intermediate_model.cc]: remove commented code?

File: caffe2/serialize/intermediate_model.h [hunk 27]
```diff
@@ -0,0 +1,399 @@
...
+    name_ = name;
+  }
+
+  void setTorchScript(const std::string& torch_script) {
+    torchScript_ = torch_script;
+  }
+
+ private:
+  std::string name_;
+  //std::unique_ptr<caffe2::NetDef> graph_;
```
Comment (dzhulgakov) [inline caffe2/serialize/intermediate_model.h]: why did you remove the NetDef? since it's forward declared it should be fine

**Review by dzhulgakov — APPROVED** [review]
remove comments and ship it?

File: caffe2/serialize/intermediate_model.h (hunk 27 above)
Comment (houseroad) [reply caffe2/serialize/intermediate_model.h]: I was doing the experiment :-)

File: caffe2/serialize/intermediate_model.cc (hunk 26 above)
Comment (houseroad) [reply caffe2/serialize/intermediate_model.cc]: Yeah, will do it

**Comment by houseroad** [comment]
@dzhulgakov I am polishing the PR, and will ship it soon :-)

**Comment by houseroad** [comment]
Use protobuf directly, so close this pr
